docs/design.md

```diff
@@ -0,0 +1,344 @@
+
+# <a name="_tr2es7dqywgj"></a>mTLS
+Mutual TLS uses a dual exchange of certificates to verify the authenticity of both the host making the gRPC call and the host receiving the request.
+
+This  requires three files on each the agent and the client:
+
+1) The CA certificate file (shared between agent and client)
+1) The host’s self signed certificate file, which is unique to the agent and each client user 
+1) The host’s private key, which is also unique to the agent and each client user
+
+gRPC uses the CA certificate file to generate a certificate pool.  The self signed certificate and the private key are used to generate the certificate which will be presented in the authentication.
+
+By default, the[ TLS package](https://go.dev/src/crypto/tls/cipher_suites.go) only allows cipher suites with no known vulnerabilities. These are the encryption algorithms that this project will use.
+
+
+# <a name="_1y2ukim9xwfx"></a>Job Service
+
+The job service is a library responsible for managing jobs. The job service provides the interface to start, stop, status and get output of jobs.
+As each job is started, it sets up the cgroup by creating the directory and setting up the appropriate files. It then manages the lifecycle of the job, both notifying the job if the client stops the job and receiving notification as jobs complete. As the job changes state, the job service updates the job repository.
+
+The lifecycle of a job:
+
+1. The cgroup mounts, directories and files are setup
+1. The command is executed using exec.cmd Go library. The process is started in a goroutine then blocks waiting for Cmd.Wait() to return. When the command is started, the pid is added to cgroup.procs file and the Cmd object is saved in the JobService. The command is called through a wrapper script which ensures that the PID of the process is added to the procs file before execution begins (see cgroups section below for details)
+1. JobService updates that job's state in the repo to running
+1. If the process is stopped prematurely via a call to Stop() the JobService kills the command.
+1. On process completion, the job's state is updated in the repo
+1. The setup done for cgroups is cleaned up.
+
+As each job is executed, the output is written to a file.  For the purposes of this project, STDOUT and STDERR are both written to the same file and are interleaved.  A more robust system would separate the outputs and allow callers to specify which should be returned.
+
+## <a name="job_service_interface">Job Service Interface
+```go
+// Start a new command.  Once started, the JobService will run the command concerntly, writing the output to a file
+// returns the command ID on success
+func (j *JobService) Start(ctx context.Context, cert *x509.Certificate, command string, args []string) (string, error)

```

Comment:
> Is the entire certificate required?

No.  It could be any field in the certificate, as long as it is unique per cert.  The `RawSubjectPublicKeyInfo` field should work too.

Reply:
> It would make more sense to me if authz/authn concerns were at a layer above the core implementation. This makes the core library more reusable if it were to be used in a way that wasn't related to gRPC/mTLS.

That's a good point.  I'll add an auth service.